### PR TITLE
1.17.x client chat received on client

### PR DESCRIPTION
**Pull request by coehlrich**
Adds an event for messages that were either received from the server or from messages that are put in chat from client-side mods or debug messages while also allowing the change the id of the message so that if a message with the same id as another id get's received then the older message get's removed.

The vanilla method that accept's the message id is currently private.

I feel like there would be a better name for the event than `ClientChatReceivedBothSidesEvent` but `ClientChatReceivedEvent` is already used which is the event that fires only for messages from the server.

### CODE REVIEW

**Comment by diesieben07** [comment]
What is the use case for this?

File: src/main/java/net/minecraftforge/client/event/ClientChatReceivedBothSidesEvent.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+import net.minecraft.network.chat.Component;
+import net.minecraftforge.eventbus.api.Cancelable;
+import net.minecraftforge.eventbus.api.Event;
+
+/**
+ * Fired when a message is received on the client from either the server or the same client
+ */
+@Cancelable
+public class ClientChatReceivedBothSidesEvent extends Event
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/client/event/ClientChatReceivedBothSidesEvent.java:30]: The javadoc as well as the name of this event do not at all accurately describe what it does.

File: src/main/java/net/minecraftforge/client/event/ClientChatReceivedBothSidesEvent.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    public int getId()
+    {
+        return id;
+    }
+
+    /*
+     * Sets the id of the message that is being received. If there is another message with the same id that was already received then that message will be removed
+     */
+    public void setId(int id)
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/client/event/ClientChatReceivedBothSidesEvent.java:62]: What is the use-case for setting the ID on a message via this event?

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 3]
```diff
@@ -149,6 +139,17 @@
...
+
+import java.io.File;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Random;
+import java.util.UUID;
+import java.util.function.Consumer;
+
 public class ForgeEventFactory
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: Do not reorganize the imports in this file.

**Review by diesieben07 — CHANGES_REQUESTED** [review]
I am really not sure what the use-case for this event would be. Please elaborate.

File: src/main/java/net/minecraftforge/client/event/ClientChatReceivedBothSidesEvent.java (hunk 1 above)
Comment (coehlrich) [reply src/main/java/net/minecraftforge/client/event/ClientChatReceivedBothSidesEvent.java:30]: I was trying to think of a better name for the event (which would probably also help with the javadoc) but couldn't think of anything other than `ClientChatReceived` which is already used for the event that only receives messages from the server.

**Comment by coehlrich** [comment]
For the use-case I was thinking it would be useful for chat message modification's that would fit for both message's coming from the server, debug keys, and other client-side mods like adding timestamps to the start of chat messages.

About the message id the use-case I had in mind was where it could be used for if a message from a server would better fit replacing the previous message of it's kind which since server's can't use the message id system it can't do that itself.

**Comment by sciwhiz12** [comment]
Hello! @coehlrich, please leave a comment whether you wish to continue this PR and update it to the latest active version of 1.18 if applicable. If we do not hear back from you after some time, this PR will be closed for inactivity.
If anyone else wishes to continue the work of this PR, please create a PR with a link to this PR for reference. 
Thank you for your contributions. 👋

**Comment by sciwhiz12** [comment]
Hello! Since the application of the Stale label, there has been no recent activity on this PR. As said in our previous comment, this PR will now be closed for inactivity.

If you wish to continue this PR, please open a new PR against the Latest supported version (to be conformant with our version support and contributions policy). 

Thank you for your contributions. 👋

**Pull request closed without merging**